datadog/resource_datadog_monitor_config_policy.go

```diff
@@ -0,0 +1,225 @@
+package datadog
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/DataDog/datadog-api-client-go/v2/api/datadogV2"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/terraform-providers/terraform-provider-datadog/datadog/internal/utils"
+	"github.com/terraform-providers/terraform-provider-datadog/datadog/internal/validators"
+)
+
+func resourceDatadogMonitorConfigPolicy() *schema.Resource {
+	return &schema.Resource{
+		Description:   "Provides a Datadog monitor config policy resource. This can be used to create and manage Datadog monitor config policies.",
+		CreateContext: resourceDatadogMonitorConfigPolicyCreate,
+		ReadContext:   resourceDatadogMonitorConfigPolicyRead,
+		UpdateContext: resourceDatadogMonitorConfigPolicyUpdate,
+		DeleteContext: resourceDatadogMonitorConfigPolicyDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"policy_type": {
+				Description:      "The monitor config policy type",
+				Type:             schema.TypeString,
+				Required:         true,
+				ValidateDiagFunc: validators.ValidateEnumValue(datadogV2.NewMonitorConfigPolicyTypeFromValue),
+			},
+			"tag_policy": {
+				Description: "Config for a tag policy. Only set if `policy_type` is `tag`.",
+				Type:        schema.TypeList,
+				Optional:    true,
+				MaxItems:    1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"tag_key": {
+							Type:        schema.TypeString,
+							Description: "The key of the tag",
+							Required:    true,
+						},
+						"tag_key_required": {
+							Type:        schema.TypeBool,
+							Description: "If a tag key is required for monitor creation",
+							Required:    true,
+						},
+						"valid_tag_values": {
+							Type:        schema.TypeList,
+							Description: "Valid values for the tag",
+							Required:    true,
+							Elem:        &schema.Schema{Type: schema.TypeString},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func buildMonitorConfigPolicyCreateV2Struct(d *schema.ResourceData) *datadogV2.MonitorConfigPolicyCreateRequest {
+	policyType, _ := datadogV2.NewMonitorConfigPolicyTypeFromValue(d.Get("policy_type").(string))
+
+	return datadogV2.NewMonitorConfigPolicyCreateRequest(
+		datadogV2.MonitorConfigPolicyCreateData{
+			Attributes: datadogV2.MonitorConfigPolicyAttributeCreateRequest{
+				PolicyType: *policyType,
+				Policy:     *buildCreateRequestPolicy(d, *policyType),
```

Comment:
Should we add a nil check here in case there was a typo? (Similarly for buildMonitorConfigPolicyUpdateV2Struct). If terraform validates the policy_type before trying to build this then I guess this isn't an issue.

Reply:
I believe the validator [here](https://github.com/DataDog/terraform-provider-datadog/pull/1750/files#diff-c39769ea0b9fa52eaa79ea763a226b2bd5e62b130c24c5806e7a17008a471d35R30) takes care of enforcing the policy_type value is in the enum set.

Reply:
Confirmed

```
resource "datadog_monitor_config_policy" "my_config_11" {
  policy_type = "test6eee"
  tag_policy {
    tag_key = "teadsfasdfasdf"
    tag_key_required = true
    valid_tag_values = ["carl", "hats", "mats"]
  }
}

$ terraform plan

│   with datadog_monitor_config_policy.my_config_11,
│   on main.tf line 36, in resource "datadog_monitor_config_policy" "my_config_11":
│   36:   policy_type = "test6eee"
│ 
│ `invalid value 'test6eee' for MonitorConfigPolicyType: valid values are [tag]`